### PR TITLE
Assert: Improve detection of bad calls to `assert.async()` callbacks

### PR DESCRIPTION
**Pull request by Krinkle**
### Background

When creating two async pauses in a test, it was possible for a test to pass by invoking one of them twice, and the other not at all.

Easy scenario (though perhaps not realistic):

> Use `assert.async()` twice, assigned as done1 and done2 in the same `QUnit.test()` case, and then simulate the failure scenario such that you wrongly call done1 two times, and forget to call done2.

Complex scenario across `QUnit.test()` and "afterEach" hooks, since these previously shared a single semaphore:

> Use `assert.async()` once in a simple test, and schedule the resume call in the future, but then fail with an uncaught error. The uncaught error is found and `Test.run()` would internally kill the pause by resetting the semaphore to zero (this make sense since we shouldn't wait for the release once the test is known to have failed).
> After this reset, we proceed to the "afterEach" hook. Suppose this hook is also async, and during its execution, the originally scheduled resume call happens. This would effectively end up releasing the afterEach's async pause despite not being finished yet, and then we proceed to the next test. That test would then fail when the afterEach's own release call happens, failing as "release during a different test".

This is the scenario of https://github.com/qunitjs/qunit/issues/1432.

Fix this and numerous other edge cases by making the returned callbacks from `assert.async()` strict about which locks they release.

Each lock now adds a unique token to a map, and invoking the release function decrements/removes this token from the map.

### Notes

* es6-map.js assigns the fallback in all browsers.
  This is a bug, to be fixed later.

* The `isNaN(semaphore)` logic was originally added in 2015 by ea3e350f.
  At the time, the internal resume function was public, and NaN could emerge through `QUnit.start("bla")` as result of `semaphore += "bla"`. This has not been possible for a while. During PR #1590, I did not trace the origin of this code, and thus did not realize that it was already obsolete (the semaphore itself is not publicly supported).

* The "during different test" error is now almost impossible to trigger since we now kill pending locks during test failures and
  tolerate all late calls equally. This meant the `drooling-done.js` test case now fails in a more limited way.

  I added a new test case for coverage, that reproduces it still, but it's a lot more obscure – it requires the original test to pass and then also have an unexpected call during a different test.

* I considered using the phrase "async lock" in the public-facing error messages, but found this perhaps too internal/technical when coming from the perspective of `var done = assert.async();`.

  In order to keep the code shared between handling of async-await, Promise, and assert.async, but remain friendly and understandable, I went for the phrase "async pause".

Fixes https://github.com/qunitjs/qunit/issues/1432.

### CODE REVIEW

File: src/test.js [hunk 1]
```diff
@@ -810,22 +811,106 @@ export function resetTestTimeout( timeoutDuration ) {
...
+//   This is close to impossible because this error only happens if the original test
+//   succesfully finished first (since other failure scenarios kill pauses and ignore
+//   late calls). It can happen if a test ended exactly as expected, but has some
+//   external or shared state continuing to hold a reference to the release callback,
+//   and either the same test scheduled another call to it in the future, or a later test
+//   causes it to be called through some shared state.
+//
+// * Pause release() is called too often, during the same test.
+//
+//   This is simply throws an error, after which uncaught error handling picks it up
```
Comment (smcclure15) [inline src/test.js:867]:
```suggestion
//   This simply throws an error, after which uncaught error handling picks it up
```

File: src/test.js [hunk 2]
```diff
@@ -810,22 +811,106 @@ export function resetTestTimeout( timeoutDuration ) {
...
+//   and processing resumes.
+export function internalStop( test, requiredCalls = 1 ) {
 	config.blocking = true;
 
-	// Set a recovery timeout, if so configured.
+	const pauseId = `#${test.asyncNextPauseId++}`;
```
Comment (smcclure15) [inline src/test.js:872]: If we're using Maps, can't we just use the integer directly as a key, instead of this stringed version? Or does that cause some memory bloat for ES5 environments?

OH. I see ahead that those are used in the messaging. This makes sense to simply store this ID once to reuse in several places later.

File: src/test.js [hunk 3]
```diff
@@ -810,22 +811,106 @@ export function resetTestTimeout( timeoutDuration ) {
...
+	const pause = {
+		killed: false,
```
Comment (smcclure15) [inline src/test.js:874]: Alternatively? `cancelled`
Coder's choice... not sure which communicates intent more clearly.

File: src/cli/run.js [hunk 4]
```diff
@@ -109,7 +109,7 @@ async function run( args, options ) {
 			console.error( "Error: Process exited before tests finished running" );
 
 			const currentTest = QUnit.config.current;
-			if ( currentTest && currentTest.semaphore ) {
+			if ( currentTest && currentTest.asyncPauses.size ) {
```
Comment (smcclure15) [inline src/cli/run.js:112]:
```suggestion
			if ( currentTest && currentTest.asyncPauses.size > 0 ) {
```
for consistency with your other checks on this

**Review by smcclure15 — APPROVED** [review]
I know my domain has some test utilities that sniff at that `semaphore` bit to get really picky about choosing to throw or continue in different scenarios. Some areas use QUnit.log and upon a failure, if they are in a pause situation, they want to jump to the next test so it's not left in a bad state. All of which I anticipate this handles much more gracefully after all.

That `semaphore` logic is completely implementation detail so that's not fair to hold up this PR, nor trigger a major version, just noting that dependency for us and others that may have found their way to relying on that.

File: src/test.js (hunk 2 above)
Comment (Krinkle) [reply src/test.js:872]: Yeah, this is a bit of a remnant. At one point I had a single map shared for all tests and even tried to cram the test name in here, too, for easier debugging, but ended up pushing it back toward the current local approach.

Looking now, I think this could even be a simple a simple array (or a Set in ES6) since we just have zero or more objects (rarely more than 1) that we push into the list and then remove (or indexOf-splice) them later. It's not like it'd be a performance issue for the small scale we have here.

TODO: Use numbers directly, move `#` to the message, same as the rest of the formatting (`[async #${id}]`).

File: src/html-reporter/es6-map.js [hunk 5]
```diff
@@ -1,14 +1,35 @@
-// Support IE 9-10, PhantomJS: Fallback for fuzzysort.js used by ./html.js
+// Support IE 9-10, Safari 7, PhantomJS: Partial Map fallback.
+// Used by html.js (via fuzzysort.js), and test.js.
+//
+// FIXME: This check is broken. This file is embedded in the qunit.js closure,
+// thus the Map var is hoisted in that scope, and starts undefined (not a function).
 var Map = typeof Map === "function" ? Map : function StringMap() {
```
Comment (gibson042) [inline src/html-reporter/es6-map.js:6]: Should we just fix it now?
```suggestion
if ( typeof Map !== "function" ) {
	var Map = function StringMap() {
```

File: src/test.js [hunk 6]
```diff
@@ -25,12 +25,13 @@ import TestReport from "./reports/test";
 export default function Test( settings ) {
 	this.expected = null;
 	this.assertions = [];
-	this.semaphore = 0;
 	this.module = config.currentModule;
 	this.steps = [];
 	this.timeout = undefined;
 	this.data = undefined;
 	this.withData = false;
+	this.asyncNextPauseId = 1;
```
Comment (gibson042) [inline src/test.js:33]: Nit: I think "asyncNextPauseId" would read more naturally as "nextAsyncPauseId" or even "asyncPausesCreated" (defaulting to 0 rather than 1).

File: src/test.js [hunk 7]
```diff
@@ -810,22 +811,106 @@ export function resetTestTimeout( timeoutDuration ) {
...
 	if ( setTimeout ) {
 		let timeoutDuration;
-
 		if ( typeof test.timeout === "number" ) {
 			timeoutDuration = test.timeout;
 		} else if ( typeof config.testTimeout === "number" ) {
 			timeoutDuration = config.testTimeout;
 		}
 
+		// Set a recovery timeout, if so configured.
```
Comment (gibson042) [inline src/test.js:913]:
```suggestion
	// Set a recovery timeout, if so configured.
	if ( setTimeout ) {
		let timeoutDuration;
		if ( typeof test.timeout === "number" ) {
			timeoutDuration = test.timeout;
		} else if ( typeof config.testTimeout === "number" ) {
			timeoutDuration = config.testTimeout;
		}

```

File: src/test.js [hunk 8]
```diff
@@ -834,8 +919,10 @@ export function internalStop( test ) {
 						`Test took longer than ${timeout}ms; test timed out.`,
 						sourceFromStacktrace( 2 )
 					);
-					released = true;
-					internalRecover( test );
+
+					pause.killed = true;
+					test.asyncPauses.delete( pauseId );
```
Comment (gibson042) [inline src/test.js:924]: Should these lines appear before invoking `pushFailure`? I'm wondering if there's any chance of `pushFailure` prompting code that can observe the not-yet-abandoned async pause.

**Review by gibson042 — COMMENTED** [review]
Nice!

File: src/html-reporter/es6-map.js (hunk 5 above)
Comment (Krinkle) [reply src/html-reporter/es6-map.js:6]: That won't work, it'll still be hosted just the same, right? It's a pretty tough situation. I set it up such that it is included by rollup as `intro` (within the file closure, so as to not leak and be seen by end-user source code), and thus the variable will be seen by fuzzysort.js and other code we use naturally as if it was a global.

Without pulling in `globalThis` into here, I'm not sure how to do this because as soon as you declare anything as `Map` you inherently also deprive your ability to see the outer scope's value for that same reference.

File: src/test.js (hunk 8 above)
Comment (Krinkle) [reply src/test.js:924]: Yep. I'll also make it call `test.pushFailure()` directly here.

File: src/html-reporter/es6-map.js (hunk 5 above)
Comment (gibson042) [reply src/html-reporter/es6-map.js:6]: Oh, right. What's needed is something like the following, which can't be fixed at this level.
```js
(function(…, Map, …) {
	if ( typeof Map !== "function" ) {
		Map = function StringMap() {…};
	}
})(…, typeof Map === "function" && Map, …)
```

File: src/test.js (hunk 7 above)
Comment (Krinkle) [reply src/test.js:913]: I didn't quite get what you meant here. I had moved the inline comment to be closer to where "if configure" happens. Are you suggesting to leave it as-is? Or is there another difference?

Comment (gibson042) [reply src/test.js:913]: I hadn't noticed it matched the preexisting pattern because of the new code, but yes. Having the inline comment just before the check for `setTimeout` inherently explains why that check is there.

File: src/test.js (hunk 6 above)
Comment (Krinkle) [reply src/test.js:33]: I had mentally namespaced these two properties under an "async" prefix, so it was meant more as "async: nextPauseId". I've dropped the prefix for now as it wasn't adding much.

File: src/test.js (hunk 7 above)
Comment (Krinkle) [reply src/test.js]: I thought of the comment as explaining that `config.timeoutHandler` happens conditionally based on whether a timeout is configured (via one of the two means). The env check for setTimeout seems unrelated, but I don't disagree that reading it up front before mentally entering the overall block is also useful. If/when the block ends up doing multiple things, then I might push to move it back, but anyway, we're not there now. I'll leave it where it was.
